gateway/pkg/tracing/otel.go

```diff
@@ -0,0 +1,155 @@
+package tracing
+
+import (
+	"context"
+	"log"
+	"os"
+	"time"
+
+	jaegerprop "go.opentelemetry.io/contrib/propagators/jaeger"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/jaeger"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+)
+
+type Exporter string
+
+const (
+	JaegerExporter   Exporter = "jaeger"
+	LogExporter      Exporter = "log"
+	OTELExporter     Exporter = "otel"
+	DisabledExporter Exporter = "disabled"
+)
+
+type Shutdown func(context.Context)
+
+// Provider returns an OpenTelemetry TracerProvider configured to use
+// the Jaeger exporter that will send spans to the provided url. The returned
+// TracerProvider will also use a Resource configured with all the information
+// about the application.
+func Provider(ctx context.Context, name, version, commit string) (shutdown Shutdown, err error) {
+	exporter := Exporter(os.Getenv("OTEL_EXPORTER"))
+
+	var exp tracesdk.TracerProviderOption
+	switch exporter {
+	case JaegerExporter:
+		// configure the collector from the env variables,
+		// OTEL_EXPORTER_JAEGER_ENDPOINT/USER/PASSWORD
+		// see: https://github.com/open-telemetry/opentelemetry-go/tree/main/exporters/jaeger
+		j, e := jaeger.New(jaeger.WithCollectorEndpoint())
+		exp, err = tracesdk.WithBatcher(j), e
+	case LogExporter:
+		w := os.Stdout
+		opts := []stdouttrace.Option{stdouttrace.WithWriter(w)}
+		if truthyEnv("OTEL_EXPORTER_LOG_PRETTY_PRINT") {
+			opts = append(opts, stdouttrace.WithPrettyPrint())
+		}
+		if !truthyEnv("OTEL_EXPORTER_LOG_TIMESTAMPS") {
+			opts = append(opts, stdouttrace.WithoutTimestamps())
+		}
+
+		s, e := stdouttrace.New(opts...)
+		exp, err = tracesdk.WithSyncer(s), e
+	case OTELExporter:
+		// find available env variables for configuration
+		// see: https://github.com/open-telemetry/opentelemetry-go/tree/main/exporters/otlp/otlptrace#environment-variables
+		kind := get("OTEL_EXPORTER_OTLP_PROTOCOL", "grpc")
```

Comment:
this is also not standard, but  i wanted a way to control which client the exporter used